### PR TITLE
schemachanger: add doc.go

### LINKED ISSUE #104779: schemachanger: document packages in a doc.go file
n/t

Jira issue: CRDB-28720

### PR DESCRIPTION
**Pull request by postamar**
Fixes #104779.

Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/doc.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+// doc file provides a high-level overview of its architecture, the aim being to
+// provide a useful introduction to the code and the design decisions behind
+// it.
+//
+// # Similarities and differences with the legacy schema changer
+//
+// The main challenge that any schema changer in CockroachDB must solve is to
+// perform these correctly while maintaining the "online" property: tables must
+// remain accessible to concurrent queries while they undergo schema changes.
+// Online schema changes are enabled by upholding the 2-version invariant, which
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:25]: should we also briefly explain what his "2-version invariant" is?

File: pkg/sql/schemachanger/doc.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+// necessarily implies that schema changes may require multiple transactions to
+// complete. Upholding correctness requires maintaining some atomicity and
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:27]: "some atomicity and isolation properties", what are they? I know it's meant to be broad but having at least one example is helpful.

File: pkg/sql/schemachanger/doc.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+// isolation properties which in turn means that schema changes which fail must
+// be rolled back somehow. Schema changes may legitimately fail, typically when
+// adding a constraint on a nonempty table.
+//
+// In CockroachDB, both in the legacy and in the declarative schema changer, for
+// schema changes involving more than one transaction. this sequence of
+// subsequent transactions is driven by a job; of type SCHEMA_CHANGE or
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:34]: get rid of "subsequent"?

File: pkg/sql/schemachanger/doc.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+// TYPEDESC_SCHEMA_CHANGE for the legacy schema changer, NEW_SCHEMA_CHANGE for
+// the declarative schema changer. In both cases the schema change gets rolled
+// back when the job's Resume method encounters an error, which eventually
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:37]: "Resume method encounters an error, after which its OnFailOrCancel method takes over and eventually transitions the job to the "failed" terminal state". 
Do you think adding those extra words in between is better?

File: pkg/sql/schemachanger/doc.go [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+// have in common the fact that they store the current state and targeted
+// end-state of an ongoing schema change inside the affected descriptor, but
+// here the similarities end: for the legacy schema changer, this state is
+// encoded in the descriptor protobuf's `mutations` slice; for the declarative
+// schema changer, it's `declarative_schema_changer_state`.
+//
+// # The element model
+//
+// The declarative schema changer models its persisted state in terms of
+// elements, which can be thought of as "things" on which a schema change can be
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:48]: "on which a schema change can effect" or "on which a schema change can be effected"?

File: pkg/sql/schemachanger/doc.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+// # Rules
+//
+// Generating these dep-edges is non-trivial and, practically speaking, requires
+// a vast quantity of complex code. This deserves further commentary.
+//
+// A graph will typically have a lot of dep-edges. Think for instance of a DROP
+// DATABASE statement operating on a database with a good many tables in it,
+// each table will easily have a few dozen elements and in such a DROP these
+// will all have target ABSENT. There will be dep-edges to ensure that, for
+// instance, a table name is not removed from system. namespace before the table
```
Comment (chengxiong-ruan) [inline pkg/sql/schemachanger/doc.go:128]: nit: remove the extra space between `system.` and `namespace`

File: pkg/sql/schemachanger/doc.go [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+//
+// # The element model
+//
+// The declarative schema changer models its persisted state in terms of
+// elements, which can be thought of as "things" on which a schema change can be
+// effected, typically leaves of a DDL statement's AST: e.g. a table's name is
+// modeled as an element called TableName, a column's default value expression
+// is modelled by ColumnDefaultExpression, and so forth.
+//
+// A schema change is define by setting target end-states to a set of elements.
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:53]: I'd suggest we be careful in our terminlogies: I tend to think it's easier if we call various states an element can be in during a schema change "status", and reserve the word "state" to be used in the context of a "schema change state" which is a collection of all targets and their current status. Similarly, a target is an element + its target status.

File: pkg/sql/schemachanger/doc.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+//
+// The advantage of this approach is that the rules can be expressed
+// using a high level of abstraction, making them easier to reason about and
+// therefore less likely to have bugs. The disadvantage is that we needed to
+// build what's essentially a mini-embedded RDBMS to apply these predicates
+// efficiently, from a performance perspective. This database is implemented in
+// `rel`.
+//
+// This is the reason why the element model is relational in nature. Each
+// element can be keyed (the keys are defined in `screl`) and therefore joined
```
Comment (chengxiong-ruan) [inline pkg/sql/schemachanger/doc.go:145]: nit: maybe say this instead: Each element can be keyed by their attributes (the attributes are defined in `screl/attr.go`

File: pkg/sql/schemachanger/doc.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+//  - ColumnName{DescID: 123, ColumnID: 4, Name: y} targets PUBLIC.
+//
+// When the schema change starts, those two elements are respectively in the
+// PUBLIC and ABSENT states. Performing the schema change means transitioning
+// these to their target states. This may include intermediate states, this
+// isn't the case for ColumnName but a Column transitions through DELETE_ONLY,
+// WRITE_ONLY, etc. because of the need to uphold the 2-version invariant.
+// Elements and states are defined in `scpb`. The mapping of the descriptor
+// model (i.e. what's in system.descriptor) to the element model (which is what
+// the declarative schema changer understands) is defined and implemented in
```
Comment (Xiang-Gu) [inline pkg/sql/schemachanger/doc.go:67]: Do you think it's helpful for the reader if we point out the fact that "one descriptor *decomposes* to a set of elements, and those elements *belong to* this one descriptor, effectively forming a one-to-many relationship?"

**Review by chengxiong-ruan — APPROVED** [review]
looks very good to me. thanks for writing this. I personally probably know most of the stuff, but never be able to describe them systematically like this.

**Review by Xiang-Gu — COMMENTED** [review]
1. somewhere in the doc, preferably in a notable place, we should explain why it's called "declarative" schema changer.

2. I think all the efforts we made into testing this framework also deserves a section. In my opinion, it's quite sophisticated and showcases the testability of the framework -- end-to-end, rollback, puase/resume, backup/restore, DML injection, three explains, and mixed version tests.

**Comment by postamar** [comment]
Those are all excellent comments, thank you very much for having given this a proper read-through!

File: pkg/sql/schemachanger/doc.go (hunk 1 above)
Comment (postamar) [reply pkg/sql/schemachanger/doc.go:25]: I'll add a link to the RFC.

File: pkg/sql/schemachanger/doc.go (hunk 2 above)
Comment (postamar) [reply pkg/sql/schemachanger/doc.go:27]: I'll explain more and provide examples.

File: pkg/sql/schemachanger/doc.go (hunk 5 above)
Comment (postamar) [reply pkg/sql/schemachanger/doc.go:48]: I've settled on "on which a schema change can be performed"

File: pkg/sql/schemachanger/doc.go (hunk 9 above)
Comment (postamar) [reply pkg/sql/schemachanger/doc.go:67]: Thanks for the suggestion but I think it's a bit of a distraction to mention that. IIRC the scdecomp and scpb packages do document that. If they don't, it belongs there.

**Comment by postamar** [comment]
I've applied almost all of the suggestions, thanks again!

bors r+

**Pull request merged**